connect/runtime/src/test/java/org/apache/kafka/connect/util/KafkaBasedLogTest.java

```diff
@@ -403,6 +404,54 @@ public void testGetOffsetsConsumerErrorOnReadToEnd() throws Exception {
         verifyStartAndStop();
     }
 
+    @Test
+    public void testOffsetReadFailureWhenWorkThreadFails() throws Exception {
+        RuntimeException exception = new RuntimeException();
+        Set<TopicPartition> tps = new HashSet<>(Arrays.asList(TP0, TP1));
+        Map<TopicPartition, Long> endOffsets = new HashMap<>();
+        endOffsets.put(TP0, 0L);
+        endOffsets.put(TP1, 0L);
+        admin = mock(TopicAdmin.class);
+        when(admin.endOffsets(eq(tps)))
+            .thenReturn(endOffsets)
+            .thenThrow(exception)
+            .thenReturn(endOffsets);
+
+        store.start();
+
+        AtomicInteger numSuccesses = new AtomicInteger();
+        AtomicInteger numFailures = new AtomicInteger();
+        AtomicReference<FutureCallback<Void>> finalSuccessCallbackRef = new AtomicReference<>();
+        final FutureCallback<Void> successCallback = new FutureCallback<>((error, result) -> numSuccesses.getAndIncrement());
+        final FutureCallback<Void> firstFailedCallback = new FutureCallback<>((error, result) -> {
+            numFailures.getAndIncrement();
+            // We issue another readToEnd call here to simulate the case that more read requests can come in while
+            // the failure is being handled in the WorkThread.
+            final FutureCallback<Void> finalSuccessCallback = new FutureCallback<>((e, r) -> numSuccesses.getAndIncrement());
+            finalSuccessCallbackRef.set(finalSuccessCallback);
+            store.readToEnd(finalSuccessCallback);
+        });
+        final FutureCallback<Void> subsequentFailedCallback = new FutureCallback<>((error, result) -> numFailures.getAndIncrement());
+
+        store.readToEnd(successCallback);
+        store.readToEnd(firstFailedCallback);
+        store.readToEnd(subsequentFailedCallback);
```

Comment:
Isn't this at risk of being flaky? Is there anything that prevents multiple callbacks from being added in rapid succession and all handled in the same iteration of the `WorkThread` loop?

Maybe instead of setting up all of our expectations for the topic admin in bulk and then issuing three consecutive calls to `readToEnd`, we could handle each case with its own expectation and call to `readToEnd` before moving on to the next?

Reply:
yeah that makes sense. I updated the test based on your suggestion.